extensions/notebook/src/book/bookUndoRedoService.ts

```diff
@@ -0,0 +1,104 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { JupyterBookSection } from '../contracts/content';
+import * as fs from 'fs-extra';
+import * as yaml from 'js-yaml';
+import { BookModel } from './bookModel';
+import * as vscode from 'vscode';
+import * as loc from '../common/localizedConstants';
+
+export interface undoRedoToc {
+	undo: JupyterBookSection[],
+	redo: JupyterBookSection[]
+}
+
+export interface IBookUndoRedoElement {
+	sourceBook: BookModel,
+	targetBook?: BookModel,
+	tocFiles: Map<string, undoRedoToc>,
+	files: Map<string, string>
+}
+
+export class BookUndoRedoService {
+	private _undoBookChange: IBookUndoRedoElement[] = [];
+	private _redoBookChange: IBookUndoRedoElement[] = [];
+	private _undoMaxSize = 10;
+
+	constructor() {
+	}
+
+	public popUndo(): IBookUndoRedoElement | undefined {
+		if (this._undoBookChange.length > 0) {
+			const change = this._undoBookChange.pop();
+			this._redoBookChange.push(change);
+			return change;
+		}
+		return undefined;
+	}
+
+	public popRedo(): IBookUndoRedoElement | undefined {
+		if (this._redoBookChange.length > 0) {
+			const change = this._redoBookChange.pop();
+			this._undoBookChange.push(change);
+			return change;
+		}
+		return undefined;
+	}
+
+	public addToUndoStack(change: IBookUndoRedoElement): void {
+		if (this._undoBookChange.length < this._undoMaxSize) {
```

Comment:
We should be removing the oldest entry when this happens - not ignoring the newest entry